README.md
# Short UUID

Generate short UUIDs or translate standard UUIDs into shorter formats
and back.

This is as direct a conversion as is possible in Dart from the node
package [short-uuid](https://github.com/oculus42/short-uuid).

## Examples

The API is very similar to the npm package with only some minor changes
required in Dart.

### Quick Start

```dart
const short = ShortUuid();

// Quick start with flickrBase58 format
short.generate(); // 73WakrfVbNJBaAmhQtEeDv
```

#### Detailed

Short UUID starts with RFC4122 v4-compliant UUIDs and translates them
into other, usually shorter formats. It also provides translators to
convert back and forth from RFC compliant UUIDs to the shorter formats.

Formats return consistent-length values unless specifically requested
otherwise. This is done by padding the start with the first ([0])
character in the alphabet. Previous versions can translate padded
formats back to UUID.

In short, it works exactly like its node counterpart.

```dart
// ShortUuid is better initiated through a call to `init` than by
// creating a new instance.
final translator = ShortUuid.init(); // Defaults to flickrBase58
final decimalTranslator = ShortUuid.init("0123456789"); // Provide a specific alphabet for translation
final cookieTranslator = ShortUuid.init(ShortUuid.constants.cookieBase90); // Use a constant for translation

// Generate a shortened v4 UUID
translator.generate(); // mhvXdrZT4jP5T8vBxuvm75

// Translate UUIDs to and from the shortened format
translator.toUUID(shortId); // a44521d0-0fb8-4ade-8002-3385545c3318
translator.fromUUID(regularUUID); // mhvXdrZT4jP5T8vBxuvm75

// Generate plain UUIDs
// - From the library without creating a translator
ShortUuid.uuid(); // fd5c084c-ff7c-4651-9a52-37096242d81c
// - Each translator provides the uuid.v4() function, too
translator.uuid(); // 3023b0f5-ec55-4e75-9cd8-104700698052

// See the alphabet used by a translator
translator.alphabet;

// View the constants
short.constants.flickrBase58; // Avoids similar characters (0/O, 1/I/l, etc.)
short.constants.cookieBase90; // Safe for HTTP cookies values for smaller IDs.
```

#### Differences

There are some slight differences between this project and the node
version ([short-uuid](https://github.com/oculus42/short-uuid)) by
oculus42.

 * Most notable is that you have to call an init function to get a
translator. This is because `ShortUuid` is a class and not a function.
 * There is no `translator.new()`. This is because `new` is a reserved
word in Dart. Instead, `translator.generate()` can be used. The
`generate` alias exists in the node version as well.

## Attributions

This is a direct conversion of Samuel Rouse's (aka oculus42) npm package
[short-uuid](https://github.com/oculus42/short-uuid) to Dart. A copy of
the project's license can be found in the [LICENSE](LICENSE) file under
the section `short-uuid`.


CHANGELOG.md
## \[2.0.0\] - 2021-05-31

* Add null safety.

## \[1.0.0\] - 2020-09-07

* Initial release.
* Functionality and API on par with `short-uuid` node project.


test/short_uuid_test.dart
import 'package:short_uuid/short_uuid.dart';
import 'package:test/test.dart';

void main() {
  test('generates a random value', () {
    final short = ShortUuid();
    final translator = ShortUuid.init();
    expect(short.generate().length, greaterThan(0));
    expect(translator.generate().length, greaterThan(0));
    expect(translator.generate().length, greaterThan(0));
  });
  test('translates to shortened format', () {
    final translator = ShortUuid.init();
    expect(translator.toUUID('mhvXdrZT4jP5T8vBxuvm75'), 'a44521d0-0fb8-4ade-8002-3385545c3318');
  });
  test('translates from shortened format', () {
    final translator = ShortUuid.init();
    expect(translator.fromUUID('a44521d0-0fb8-4ade-8002-3385545c3318'), 'mhvXdrZT4jP5T8vBxuvm75');
  });
}


test/short_uuids_test.dart
import 'package:short_uuids/short_uuids.dart';
import 'package:test/test.dart';

void main() {
  test('generates a random value', () {
    final short = ShortUuid();
    final translator = ShortUuid.init();
    expect(short.generate().length, greaterThan(0));
    expect(translator.generate().length, greaterThan(0));
    expect(translator.generate().length, greaterThan(0));
  });
  test('translates to shortened format', () {
    final translator = ShortUuid.init();
    expect(translator.toUUID('mhvXdrZT4jP5T8vBxuvm75'), 'a44521d0-0fb8-4ade-8002-3385545c3318');
  });
  test('translates from shortened format', () {
    final translator = ShortUuid.init();
    expect(translator.fromUUID('a44521d0-0fb8-4ade-8002-3385545c3318'), 'mhvXdrZT4jP5T8vBxuvm75');
  });
}


lib/short_uuids.dart
library short_uuid;

import 'dart:math';

import 'package:any_base/any_base.dart';
import 'package:uuid/uuid.dart';

typedef Generator = String Function();
typedef Translator = String Function(String);

class PaddingParams {
  final bool consistentLength;
  final int shortIdLength;
  final paddingChar;
  const PaddingParams({
    this.consistentLength = true,
    required this.shortIdLength,
    this.paddingChar = '0',
  });
}

class Convertor {
  final Generator generate;
  final Generator uuid;
  final Translator fromUUID;
  final Translator toUUID;
  final String alphabet;
  const Convertor({
    required this.generate,
    required this.uuid,
    required this.fromUUID,
    required this.toUUID,
    required this.alphabet
  });
}

class _Constants {
  final flickrBase58 = '123456789abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ';
  final cookieBase90 = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!#\$%&'()*+-./:<=>?@[]^_`{|}~";
  const _Constants();
}

class ShortUuid {
  static const constants = _Constants();
  static final uuidV4 = Uuid().v4;

  // A default generator, instantiated only if used.
  static Convertor? _toFlickr;

  // Expose the UUIDv4 generator.
  static String Function({Map<String, dynamic> options}) get uuid => uuidV4;

  /// Initialize the default translator.
  static Convertor init([String? toAlphabet, PaddingParams? options]) =>
      makeConvertor(toAlphabet, options);

  const ShortUuid();

  /// Takes a UUID, strips the dashes, and translates.
  ///
  /// Strips [longId] of dashes and passes it to the [translator] function to be
  /// converted into shortened form and returns the result.
  ///
  /// Adds padding in front of the resulting short form if [paddingParams] has
  /// `consistentLength` set to true.
  static String shortenUUID(String longId, Translator translator,
      [PaddingParams? paddingParams]) {
    final translated = translator(longId.toLowerCase().replaceAll('-', ''));

    if (paddingParams == null || !paddingParams.consistentLength)
      return translated;

    return translated.padLeft(
      paddingParams.shortIdLength,
      paddingParams.paddingChar,
    );
  }

  /// Translate back to hex and turn back into UUID format, adding dashes.
  ///
  /// Passes [shortId] to the [translator] function to be expanded to hex.
  /// Dashes are then added to construct a standard uuid string and the
  /// resulting `String` is returned.
  static String enlargeUUID(String shortId, Translator translator) {
    final uu1 = translator(shortId).padLeft(32, '0');

    // Join the zero padding and the UUID and then slice it up with match
    final m = RegExp(r'(\w{8})(\w{4})(\w{4})(\w{4})(\w{12})').firstMatch(uu1);

    // Accumulate the matches and join them.
    return [m![1], m[2], m[3], m[4], m[5]].join('-');
  }

  /// Calculate the length of the shortened id.
  static int getShortIdLength(alphabetLength) =>
      (log(pow(2.0, 128.0)) / log(alphabetLength.toDouble())).ceil();

  /// Create a custom convertor using a specified alphabet.
  static Convertor makeConvertor([String? toAlphabet, PaddingParams? options]) {
    // Default to Flickr 58
    final useAlphabet = toAlphabet ?? constants.flickrBase58;

    // Default to baseOptions
    final selectedOptions = options ?? PaddingParams(
      shortIdLength: getShortIdLength(useAlphabet.length),
      paddingChar: useAlphabet[0],
    );

    // Check alphabet for duplicate entries
    if (Set.from(useAlphabet.split('')).length != useAlphabet.length) {
      throw FormatException(
          'The provided Alphabet has duplicate characters resulting in unreliable results');
    }

    // UUIDs are in hex, so we translate to and from.
    final anyBase = AnyBase(AnyBase.hex, useAlphabet);
    final fromHex = anyBase.convert;
    final toHex = anyBase.revert;
    final generate = () => shortenUUID(uuidV4(), fromHex, options);

    return Convertor(
      generate: generate,
      uuid: uuidV4,
      fromUUID: (uuid) => shortenUUID(uuid, fromHex, selectedOptions),
      toUUID: (shortUuid) => enlargeUUID(shortUuid, toHex),
      alphabet: useAlphabet,
    );
  }

  // Provide a generic generator
  String generate() {
    if (_toFlickr == null) {
      // Generate on first use;
      _toFlickr = makeConvertor(constants.flickrBase58);
    }
    return shortenUUID(uuid(), _toFlickr!.fromUUID);
  }
}


lib/short_uuid.dart
library short_uuid;

import 'dart:math';

import 'package:any_base/any_base.dart';
import 'package:uuid/uuid.dart';

typedef Generator = String Function();
typedef Translator = String Function(String);

class PaddingParams {
  final bool consistentLength;
  final int shortIdLength;
  final paddingChar;
  const PaddingParams({
    this.consistentLength = true,
    this.shortIdLength,
    this.paddingChar = '0',
  }) : assert(consistentLength != null),
       assert(shortIdLength != null),
       assert(paddingChar != null);
}

class Convertor {
  final Generator generate;
  final Generator uuid;
  final Translator fromUUID;
  final Translator toUUID;
  final String alphabet;
  const Convertor({
    this.generate, this.uuid, this.fromUUID, this.toUUID, this.alphabet
  })  : assert(generate != null),
        assert(uuid != null),
        assert(fromUUID != null),
        assert(toUUID != null),
        assert(alphabet != null);
}

class _Constants {
  final flickrBase58 = '123456789abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ';
  final cookieBase90 = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!#\$%&'()*+-./:<=>?@[]^_`{|}~";
  const _Constants();
}

class ShortUuid {
  static const constants = _Constants();
  static final uuidV4 = Uuid().v4;

  // A default generator, instantiated only if used.
  static Convertor _toFlickr;

  // Expose the UUIDv4 generator.
  static String Function({Map<String, dynamic> options}) get uuid => uuidV4;

  /// Initialize the default translator.
  static Convertor init([String toAlphabet, PaddingParams options]) =>
      makeConvertor(toAlphabet, options);

  const ShortUuid();

  /// Takes a UUID, strips the dashes, and translates.
  ///
  /// Strips [longId] of dashes and passes it to the [translator] function to be
  /// converted into shortened form and returns the result.
  ///
  /// Adds padding in front of the resulting short form if [paddingParams] has
  /// `consistentLength` set to true.
  static String shortenUUID(String longId, Translator translator,
      [PaddingParams paddingParams]) {
    final translated = translator(longId.toLowerCase().replaceAll('-', ''));

    if (paddingParams == null || !paddingParams.consistentLength)
      return translated;

    return translated.padLeft(
      paddingParams.shortIdLength,
      paddingParams.paddingChar,
    );
  }

  /// Translate back to hex and turn back into UUID format, adding dashes.
  ///
  /// Passes [shortId] to the [translator] function to be expanded to hex.
  /// Dashes are then added to construct a standard uuid string and the
  /// resulting `String` is returned.
  static String enlargeUUID(String shortId, Translator translator) {
    final uu1 = translator(shortId).padLeft(32, '0');

    // Join the zero padding and the UUID and then slice it up with match
    final m = RegExp(r'(\w{8})(\w{4})(\w{4})(\w{4})(\w{12})').firstMatch(uu1);

    // Accumulate the matches and join them.
    return [m[1], m[2], m[3], m[4], m[5]].join('-');
  }

  /// Calculate the length of the shortened id.
  static int getShortIdLength(alphabetLength) =>
      (log(pow(2.0, 128.0)) / log(alphabetLength.toDouble())).ceil();

  /// Create a custom convertor using a specified alphabet.
  static Convertor makeConvertor([String toAlphabet, PaddingParams options]) {
    // Default to Flickr 58
    final useAlphabet = toAlphabet ?? constants.flickrBase58;

    // Default to baseOptions
    final selectedOptions = options ?? PaddingParams(
      shortIdLength: getShortIdLength(useAlphabet.length),
      paddingChar: useAlphabet[0],
    );

    // Check alphabet for duplicate entries
    if (Set.from(useAlphabet.split('')).length != useAlphabet.length) {
      throw FormatException(
          'The provided Alphabet has duplicate characters resulting in unreliable results');
    }

    // UUIDs are in hex, so we translate to and from.
    final anyBase = AnyBase(AnyBase.hex, useAlphabet);
    final fromHex = anyBase.convert;
    final toHex = anyBase.revert;
    final generate = () => shortenUUID(uuidV4(), fromHex, options);

    return Convertor(
      generate: generate,
      uuid: uuidV4,
      fromUUID: (uuid) => shortenUUID(uuid, fromHex, selectedOptions),
      toUUID: (shortUuid) => enlargeUUID(shortUuid, toHex),
      alphabet: useAlphabet,
    );
  }

  // Provide a generic generator
  String generate() {
    if (_toFlickr == null) {
      // Generate on first use;
      _toFlickr = makeConvertor(constants.flickrBase58);
    }
    return shortenUUID(uuid(), _toFlickr.fromUUID);
  }
}
